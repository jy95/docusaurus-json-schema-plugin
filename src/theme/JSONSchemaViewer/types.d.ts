import type { JSONSchema as Draft_07 } from "json-schema-typed/draft-07"
import type { JSONSchema as Draft_2019_09 } from "json-schema-typed/draft-2019-09"
import type { JSONSchema as Draft_2020_12 } from "json-schema-typed/draft-2020-12"

export type JSONSchema = Draft_07 | Draft_2019_09 | Draft_2020_12

// In case I want to cover one specific file
export type JSONSchema_Draft_07 = Draft_07
export type JSONSchema_Draft_2019_09 = Draft_2019_09
export type JSONSchema_Draft_2020_12 = Draft_2020_12

// Namespace, In case I need to cast to a specific type (array / object)
import type { JSONSchema as JSONSchemaNS } from "json-schema-typed"
export { JSONSchemaNS }

export type { TypeName as TypeName } from "json-schema-typed"
export type { keywords } from "json-schema-typed"
