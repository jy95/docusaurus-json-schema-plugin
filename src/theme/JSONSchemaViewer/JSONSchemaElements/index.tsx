// Component to deal with oneOf / anyOf / allOf / not
export { SchemaComposition } from "@theme/JSONSchemaViewer/JSONSchemaElements/schemaComposition"

// Component to deal with if-then-else , dependentRequired , dependentSchemas , dependencies
export { SchemaConditional } from "@theme/JSONSchemaViewer/JSONSchemaElements/SchemaConditional"

// Component to deal with Object
export { CreateObject } from "@theme/JSONSchemaViewer/JSONSchemaElements/object"

// Component to deal with Array
export { CreateArray } from "@theme/JSONSchemaViewer/JSONSchemaElements/array"

// Component to deal with String
export { default as CreateString } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateString"

// Component to deal with Boolean
export { default as CreateBoolean } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateBoolean"

// Component to deal with Number
export { default as CreateNumber } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateNumber"

// Component to deal with Integer
export { default as CreateInteger } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateInteger"

// Component to deal with null
export { default as CreateNull } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateNull"

// Component to deal with schema defined as "true"
export { default as CreateAlwaysValid } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateAlwaysValid"

// Component to deal with schema defined as "false"
export { default as CreateAlwaysInvalid } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateAlwaysInvalid"

// Component to deal with description
export { default as CreateDescription } from "@theme/JSONSchemaViewer/JSONSchemaElements/CreateDescription"
