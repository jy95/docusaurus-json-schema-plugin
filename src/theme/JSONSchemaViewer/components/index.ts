// To avoid circular dependancies issue, let's have a single place for exports and imports

export { default as Collapsible } from "./Collapsible"

export { default as CreateNodes } from "./CreateNodes"

export { default as CreateEdge } from "./CreateEdge"

export { default as SchemaItem } from "./SchemaItem/SchemaItem"
