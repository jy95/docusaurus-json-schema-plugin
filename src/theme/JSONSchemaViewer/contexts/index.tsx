export * from "@theme/JSONSchemaViewer/contexts/jsvOptions"
