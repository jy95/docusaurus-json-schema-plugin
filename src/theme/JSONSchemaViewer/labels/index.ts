// Re-export labels
export { default as ObjectLabel } from "@theme/JSONSchemaViewer/labels/Object"
export { default as ArrayLabel } from "@theme/JSONSchemaViewer/labels/Array"
export { default as TypeLabel } from "@theme/JSONSchemaViewer/labels/Type"
export { default as IfLabel } from "@theme/JSONSchemaViewer/labels/If"
export { default as ThenLabel } from "@theme/JSONSchemaViewer/labels/Then"
export { default as ElseLabel } from "@theme/JSONSchemaViewer/labels/Else"
export { default as LoadingLabel } from "@theme/JSONSchemaViewer/labels/Loading"
export { default as ErrorOccurredLabel } from "@theme/JSONSchemaViewer/labels/ErrorOccurred"
export { default as RequiredLabel } from "@theme/JSONSchemaViewer/labels/Required/Required"
export { default as DeprecatedLabel } from "@theme/JSONSchemaViewer/labels/Deprecated/Deprecated"
export { default as ReadOnlyLabel } from "@theme/JSONSchemaViewer/labels/ReadOnly/ReadOnly"
export { default as WriteOnlyLabel } from "@theme/JSONSchemaViewer/labels/WriteOnly/WriteOnly"
export { default as TrueLabel } from "@theme/JSONSchemaViewer/labels/True"
export { default as FalseLabel } from "@theme/JSONSchemaViewer/labels/False"
export { default as StringLabel } from "@theme/JSONSchemaViewer/labels/String"
export { default as BooleanLabel } from "@theme/JSONSchemaViewer/labels/Boolean"
export { default as NumberLabel } from "@theme/JSONSchemaViewer/labels/Number"
export { default as IntegerLabel } from "@theme/JSONSchemaViewer/labels/Integer"
export { default as NullLabel } from "@theme/JSONSchemaViewer/labels/Null"
export { default as AndLabel } from "@theme/JSONSchemaViewer/labels/And"
export { default as OrLabel } from "@theme/JSONSchemaViewer/labels/Or"
export { default as NotLabel } from "@theme/JSONSchemaViewer/labels/Not"
export { default as XorLabel } from "@theme/JSONSchemaViewer/labels/Xor"
