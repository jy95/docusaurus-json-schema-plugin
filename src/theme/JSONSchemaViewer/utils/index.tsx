// To keep my mind happy, let's have a place where I can import stuff without knowing their final place in internal structure

// Re export generateFriendlyName function
export { default as GenerateFriendlyName } from "@theme/JSONSchemaViewer/utils/generateFriendlyName"

// Re export QualifierMessages function
export { default as QualifierMessages } from "@theme/JSONSchemaViewer/utils/getQualifierMessages"

// Re export detectTypes functions
export * from "@theme/JSONSchemaViewer/utils/detectTypes"
