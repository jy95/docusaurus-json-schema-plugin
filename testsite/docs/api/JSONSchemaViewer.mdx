---
title: "ðŸ”Ž @theme/JSONSchemaViewer"
description: "@theme/JSONSchemaViewer API"
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

## Configuration

| Property        | Type          | Required ?  | Note                                                    |
|-----------------|---------------|-------------|---------------------------------------------------------|
| schema          | JSONSchema    | Mandatory   | JSON Schema Draft-07 / Draft 2019-09 / Draft 2020-12    |
| resolverOptions | IResolveOpts  | Optional    | To resolve your $ref (by default, only inline references will be dereferenced). |
| viewerOptions   | JSVOptions    | Optional    | Options for the viewer itself. |

### Types

```tsx
// Either Draft-07 / Draft 2019-09 / Draft 2020-12
type JSONSchema = unknown
```

```tsx
type IResolveOpts = {
    // "IResolveOpts" options from @stoplightio/json-ref-resolver
    // More info on https://github.com/stoplightio/json-ref-resolver
    // https://github.com/stoplightio/json-ref-resolver/blob/master/src/types.ts
}
```

```tsx
type JSVOptions = {
   /**
   * Should we display "examples" ?
   * @default false
   */
    showExamples?: boolean
}
```

## Examples

<Tabs>
<TabItem value="JSX" label="JSX">

```tsx
import React from "react"
import Layout from "@theme/Layout"
import JSONSchemaViewer from "@theme/JSONSchemaViewer"

export default function ExamplePage(): JSX.Element {

    // You are free to fetch your schema in your own way (load local file, fetch, ...) :)
    const mySchema = {
        "type": "object",
        "properties": {
            "builtin": {
                "type": "number"
            }
        },
        "patternProperties": {
            "^S_": {
                "type": "string"
            },
            "^I_": {
                "type": "integer"
            }
        },
        "additionalProperties": {
            "type": "string"
        }
    }

    return (
        <Layout
            title={`My super JSON Schema`}
            description="Description will go into a meta tag in <head />"
        >
            <JSONSchemaViewer schema={mySchema} />
        </Layout>
    )
}
```

</TabItem>

<TabItem value="MDX" label="MDX" default>

```tsx

import CodeBlock from '@theme/CodeBlock';
// To fetch a JSON file from your static folder
import Schema from "@site/static/schemas/examples/array/tuples.json";
import JSONSchemaViewer from "@theme/JSONSchemaViewer";

# Tuples

Viewer :

<JSONSchemaViewer schema={ Schema  } />

Source :

<CodeBlock language="json">{JSON.stringify(Schema, null, 2)}</CodeBlock>

```

</TabItem>

</Tabs>