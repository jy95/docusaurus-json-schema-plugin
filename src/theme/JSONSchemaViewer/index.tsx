import React from "react"
import { Resolver } from "@stoplight/json-ref-resolver"

import { CreateNodes, Collapsible } from "./components/index"

import type { JSONSchema7 } from "json-schema"
import type { IResolverOpts } from "@stoplight/json-ref-resolver/types"

export type Props = {
  schema: JSONSchema7
  // To customize the ref resolving
  resolverOptions?: IResolverOpts
  [x: string]: any
}

type InnerViewerProperties = {
  // Thanks to @stoplight/json-ref-resolver, $ref are either :
  // 1. resolved
  // 2. unresolved (as circular stuff are not on the roadmap)
  schema: JSONSchema7
}

// Internal
function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  const { schema } = props
  // Title of the schema, for user friendliness
  const title = schema?.title || "Schema"
  return (
    <Collapsible
      summary={<strong>{title}</strong>}
      detailsProps={{
        open: true,
      }}
    >
      <CreateNodes schema={schema} />
    </Collapsible>
  )
}

// Entry point
function JSONSchemaViewer(props: Props): JSX.Element {
  const { schema: originalSchema, resolverOptions } = props

  // Set up resolver
  // @ts-ignore
  const resolver = new Resolver(resolverOptions)

  // TODO How do I wait for async to be complete before I run that ?
  // resolver.resolve(originalSchema);

  // Simplify schema
  //const simplifiedSchema = (originalSchema?.allOf !== undefined) ? mergeAllOf(originalSchema) : originalSchema;

  return <JSONSchemaInnerViewer schema={originalSchema} />
}

export default JSONSchemaViewer
