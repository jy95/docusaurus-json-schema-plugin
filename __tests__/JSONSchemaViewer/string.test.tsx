import React from "react"
import { render, act } from "@testing-library/react"
import { expect, test, describe } from "@jest/globals"
import JSONSchemaViewer from "../../src/theme/JSONSchemaViewer/index"

// Type to prevent creating invalid mocks
import type { RenderResult } from "@testing-library/react"
import type { JSONSchema } from "../../src/theme/JSONSchemaViewer/types"

const testcases: [string, JSONSchema][] = [
  [
    "basic",
    {
      type: "string",
    },
  ],
  [
    "format",
    {
      type: "object",
      properties: {
        prop1: {
          type: "string",
          format: "date-time",
          examples: ["2018-11-13T20:20:39+00:00"],
        },
        prop2: {
          type: "string",
          format: "time",
          examples: ["20:20:39+00:00"],
        },
        prop3: {
          type: "string",
          format: "date",
          examples: ["2018-11-13"],
        },
        prop4: {
          type: "string",
          format: "duration",
          examples: ["P3D"],
        },
        prop5: {
          type: "string",
          format: "email",
          examples: ["test@gmail.com"],
        },
        prop6: {
          type: "string",
          format: "idn-email",
          examples: ["test@gmail.com"],
        },
        prop7: {
          type: "string",
          format: "hostname",
          examples: ["example"],
        },
        prop8: {
          type: "string",
          format: "idn-hostname",
          examples: ["example"],
        },
        prop9: {
          type: "string",
          format: "ipv4",
          examples: ["192.168.1.1"],
        },
        prop10: {
          type: "string",
          format: "ipv6",
          examples: ["2001:db8:3333:4444:5555:6666:7777:8888"],
        },
        prop11: {
          type: "string",
          format: "uuid",
          examples: ["3e4666bf-d5e5-4aa7-b8ce-cefe41c7568a"],
        },
        prop12: {
          type: "string",
          format: "uri",
          examples: ["https://www.perdu.com/"],
        },
        prop12B: {
          type: "string",
          format: "uri-reference",
        },
        prop13: {
          type: "string",
          format: "iri",
          examples: ["https://www.perdu.com/"],
        },
        prop13B: {
          type: "string",
          format: "iri-reference",
        },
        prop14: {
          type: "string",
          format: "uri-template",
        },
        prop15: {
          type: "string",
          format: "json-pointer",
        },
        prop16: {
          type: "string",
          format: "relative-json-pointer",
        },
        prop17: {
          type: "string",
          format: "regex",
          examples: ["^(\\([0-9]{3}\\))?[0-9]{3}-[0-9]{4}$"],
        },
      },
    },
  ],
  [
    "minLength",
    {
      type: "string",
      minLength: 2,
    },
  ],
  [
    "maxLength",
    {
      type: "string",
      maxLength: 3,
    },
  ],
  [
    "minLength & maxLength",
    {
      type: "string",
      minLength: 2,
      maxLength: 3,
    },
  ],
  [
    "pattern",
    {
      type: "string",
      pattern: "^(\\([0-9]{3}\\))?[0-9]{3}-[0-9]{4}$",
    },
  ],
  [
    "contentEncoding",
    {
      type: "string",
      contentEncoding: "base64",
    },
  ],
  [
    "contentMediaType",
    {
      type: "string",
      contentMediaType: "application/json",
    },
  ],
  [
    "contentEncoding with contentMediaType",
    {
      type: "string",
      contentEncoding: "base64",
      contentMediaType: "application/json",
    },
  ],
  [
    "contentSchema",
    {
      type: "string",
      contentSchema: {
        type: "object",
        required: ["name", "age"],
      },
    },
  ],
]

describe("JSONSchemaViewer - string", () => {
  test.each(testcases)("test %s", async (title, fakeSchema) => {
    // Render the component
    let rendered: RenderResult | null = null

    // Use act to ensure all updates are processed
    await act(async () => {
      rendered = render(<JSONSchemaViewer schema={fakeSchema} />)
    })

    // Capture the snapshot
    expect(rendered!.asFragment()).toMatchSnapshot()
  })
})
