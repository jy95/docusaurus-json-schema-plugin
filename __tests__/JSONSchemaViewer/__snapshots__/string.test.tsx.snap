// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSchemaViewer - string test basic 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
         : 
        <span
          style="opacity: 0.6;"
        >
          string
        </span>
        <div />
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - string test contentEncoding 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
         : 
        <span
          style="opacity: 0.6;"
        >
          string
        </span>
        <div>
          <div>
            <strong>
              Encoding :
            </strong>
             
            <code>
              base64
            </code>
          </div>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - string test contentEncoding with contentMediaType 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
         : 
        <span
          style="opacity: 0.6;"
        >
          string
        </span>
        <div>
          <div>
            <strong>
              Encoding :
            </strong>
             
            <code>
              base64
            </code>
          </div>
          <div>
            <strong>
              Media type :
            </strong>
             
            <code>
              application/json
            </code>
          </div>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - string test contentMediaType 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
         : 
        <span
          style="opacity: 0.6;"
        >
          string
        </span>
        <div>
          <div>
            <strong>
              Media type :
            </strong>
             
            <code>
              application/json
            </code>
          </div>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - string test contentSchema 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
         : 
        <span
          style="opacity: 0.6;"
        >
          string
        </span>
        <div>
          <div>
            <strong>
              Decoded content must be validated against this schema :
            </strong>
             
            <details
              open=""
            >
              <summary>
                <strong>
                  Schema
                </strong>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    object
                  </span>
                  <ul>
                    <li
                      data-level="2"
                      id="/contentSchema/properties/name"
                    >
                      <details>
                        <summary>
                          <strong>
                            name
                          </strong>
                           
                          <span
                            style="opacity: 0.6;"
                          >
                            any
                          </span>
                           
                          <strong>
                            required
                          </strong>
                        </summary>
                        <div>
                          <div>
                            <strong>
                              type
                            </strong>
                             : 
                            <span
                              style="opacity: 0.6;"
                            >
                              any
                            </span>
                            <div>
                              Always valid
                            </div>
                          </div>
                        </div>
                      </details>
                    </li>
                    <li
                      data-level="2"
                      id="/contentSchema/properties/age"
                    >
                      <details>
                        <summary>
                          <strong>
                            age
                          </strong>
                           
                          <span
                            style="opacity: 0.6;"
                          >
                            any
                          </span>
                           
                          <strong>
                            required
                          </strong>
                        </summary>
                        <div>
                          <div>
                            <strong>
                              type
                            </strong>
                             : 
                            <span
                              style="opacity: 0.6;"
                            >
                              any
                            </span>
                            <div>
                              Always valid
                            </div>
                          </div>
                        </div>
                      </details>
                    </li>
                  </ul>
                  <div />
                </div>
              </div>
            </details>
          </div>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - string test format 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
         : 
        <span
          style="opacity: 0.6;"
        >
          object
        </span>
        <ul>
          <li
            data-level="1"
            id="/properties/prop1"
          >
            <details>
              <summary>
                <strong>
                  prop1
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  date-time
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (date-time)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop2"
          >
            <details>
              <summary>
                <strong>
                  prop2
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  time
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (time)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop3"
          >
            <details>
              <summary>
                <strong>
                  prop3
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  date
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (date)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop4"
          >
            <details>
              <summary>
                <strong>
                  prop4
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  duration
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (duration)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop5"
          >
            <details>
              <summary>
                <strong>
                  prop5
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  email
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (email)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop6"
          >
            <details>
              <summary>
                <strong>
                  prop6
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  idn-email
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (idn-email)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop7"
          >
            <details>
              <summary>
                <strong>
                  prop7
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  hostname
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (hostname)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop8"
          >
            <details>
              <summary>
                <strong>
                  prop8
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  idn-hostname
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (idn-hostname)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop9"
          >
            <details>
              <summary>
                <strong>
                  prop9
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  ipv4
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (ipv4)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop10"
          >
            <details>
              <summary>
                <strong>
                  prop10
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  ipv6
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (ipv6)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop11"
          >
            <details>
              <summary>
                <strong>
                  prop11
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  uuid
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (uuid)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop12"
          >
            <details>
              <summary>
                <strong>
                  prop12
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  uri
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (uri)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop12B"
          >
            <details>
              <summary>
                <strong>
                  prop12B
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  uri-reference
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (uri-reference)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop13"
          >
            <details>
              <summary>
                <strong>
                  prop13
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  iri
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (iri)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop13B"
          >
            <details>
              <summary>
                <strong>
                  prop13B
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  iri-reference
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (iri-reference)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop14"
          >
            <details>
              <summary>
                <strong>
                  prop14
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  uri-template
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (uri-template)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop15"
          >
            <details>
              <summary>
                <strong>
                  prop15
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  json-pointer
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (json-pointer)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop16"
          >
            <details>
              <summary>
                <strong>
                  prop16
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  relative-json-pointer
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (relative-json-pointer)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
          <li
            data-level="1"
            id="/properties/prop17"
          >
            <details>
              <summary>
                <strong>
                  prop17
                </strong>
                 
                <span
                  style="opacity: 0.6;"
                >
                  regex
                </span>
              </summary>
              <div>
                <div>
                  <strong>
                    type
                  </strong>
                   : 
                  <span
                    style="opacity: 0.6;"
                  >
                    string (regex)
                  </span>
                  <div />
                </div>
              </div>
            </details>
          </li>
        </ul>
        <div />
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - string test maxLength 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
         : 
        <span
          style="opacity: 0.6;"
        >
          string
        </span>
        <div>
          <div>
            <strong>
              Length :
            </strong>
             
            <code>
              &lt;= 3 character(s)
            </code>
          </div>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - string test minLength & maxLength 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
         : 
        <span
          style="opacity: 0.6;"
        >
          string
        </span>
        <div>
          <div>
            <strong>
              Length :
            </strong>
             
            <code>
              &gt;= 2 character(s)
            </code>
             
            <strong>
              AND
            </strong>
             
            <code>
              &lt;= 3 character(s)
            </code>
          </div>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - string test minLength 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
         : 
        <span
          style="opacity: 0.6;"
        >
          string
        </span>
        <div>
          <div>
            <strong>
              Length :
            </strong>
             
            <code>
              &gt;= 2 character(s)
            </code>
          </div>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;

exports[`JSONSchemaViewer - string test pattern 1`] = `
<DocumentFragment>
  <details
    class="json-schema-viewer"
    open=""
  >
    <summary>
      <strong>
        Schema
      </strong>
    </summary>
    <div>
      <div>
        <strong>
          type
        </strong>
         : 
        <span
          style="opacity: 0.6;"
        >
          string
        </span>
        <div>
          <div>
            <strong>
              Pattern :
            </strong>
             
            <code>
              ^(\\([0-9]{3}\\))?[0-9]{3}-[0-9]{4}$
            </code>
          </div>
        </div>
      </div>
    </div>
  </details>
</DocumentFragment>
`;
